File: commons/src/main/java/com/icostel/commons/utils/prefs/PersistentSetting.java
package com.icostel.commons.utils.prefs;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.Gson;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import timber.log.Timber;

public abstract class PersistentSetting<T> implements Setting<T> {
    private SharedPreferences preferences;
    private PreferenceLiveData data;
    private Gson gson;

    public PersistentSetting(SharedPreferences preferences) {
        this.preferences = preferences;
        gson = new Gson();

        data = new PreferenceLiveData();
        data.postValue(read());
    }

    @Override
    public LiveData<T> value() {
        return data;
    }

    @Override
    public T get() {
        return data.getValue();
    }

    @Override
    public T instantValue() {
        T instantValue = read();
        data.postValue(instantValue);
        return instantValue;
    }

    @Override
    public void updateValue(T value) {
        write(value);
        data.postValue(value);
    }

    @Override
    public boolean hasValue() {
        return data.getValue() != null;
    }

    @SuppressWarnings("unchecked")
    private T read() {
        try {
            Class<T> clazz = clazz();
            if (clazz == String.class) {
                return (T) preferences.getString(key(), null);
            } else if (clazz == Boolean.class) {
                return (T) Boolean.valueOf(preferences.getString(key(), null));
            } else if (clazz == Integer.class) {
                return (T) Integer.valueOf(preferences.getString(key(), null));
            } else {
                // Fallback to Gson
                return gson.fromJson(preferences.getString(key(), null), clazz);
            }
        } catch (Exception e) {
            Timber.e(e);
        }
        return null;
    }

    private void write(T data) {
        Class<T> clazz = clazz();
        String encoded = null;
        if (data != null) {
            if (clazz == String.class) {
                encoded = (String) data;
            } else if (clazz == Boolean.class) {
                encoded = String.valueOf(data);
            } else if (clazz == Integer.class) {
                encoded = String.valueOf(data);
            } else {
                // Fallback to Gson
                encoded = gson.toJson(data);
            }
        }
        Timber.d("write() " + encoded + " with key: " + key());
        preferences.edit().putString(key(), encoded).apply();
        this.data.postValue(data);
    }

    private void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (TextUtils.equals(key, key())) {
            data.setValue(read());
        }
    }

    protected abstract String key();

    protected abstract Class<T> clazz();

    private class PreferenceLiveData extends MutableLiveData<T> {
        @Override
        protected void onActive() {
            super.onActive();
            preferences.registerOnSharedPreferenceChangeListener(PersistentSetting.this::onSharedPreferenceChanged);
        }

        @Override
        protected void onInactive() {
            super.onInactive();
            preferences.unregisterOnSharedPreferenceChangeListener(PersistentSetting.this::onSharedPreferenceChanged);
        }
    }
}

File: commons/src/main/java/com/icostel/commons/utils/prefs/PersistentSetting.kt
package com.icostel.commons.utils.prefs

import android.content.SharedPreferences
import android.text.TextUtils

import com.google.gson.Gson

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import timber.log.Timber

abstract class PersistentSetting<T>(
        private val preferences: SharedPreferences,
        private val gSon: Gson = Gson()
) : Setting<T> {

    private val data: PreferenceLiveData = PreferenceLiveData()

    //register to data events from here
    override fun getData(): LiveData<T?> {
        return data
    }

    override // gets the latest getData stored and pushed a new update of the data
    fun getValue(): T? {
        val instantValue = read()
        data.postValue(instantValue)
        return instantValue
    }

    override fun updateValue(value: T?) {
        write(value)
        data.postValue(value)
    }

    override fun hasValue(): Boolean {
        return data.value != null
    }

    @Suppress("UNCHECKED_CAST")
    private fun read(): T? {
        try {
            val clazz = clazz()
            return when(clazz()) {
                String::class.java -> preferences.getString(key(), null) as T?
                Boolean::class.java -> preferences.getString(key(), null)?.toBoolean() as T?
                Int::class.java -> preferences.getString(key(), null)?.toInt() as T?
                else -> {
                    gSon.fromJson(preferences.getString(key(), null), clazz)
                }
            }
        } catch (e: Exception) {
            Timber.e(e)
        }

        return null
    }

    private fun write(data: T?) {
        if (data == null) {
            preferences.edit().remove(key()).apply()
        } else {
            val encoded = when(clazz()) {
                String::class.java -> data as String?
                Boolean::class.java, Int::class.java -> data.toString()
                else -> gSon.toJson(data)
            }
            preferences.edit().putString(key(), encoded).apply()
        }
        this.data.postValue(data)
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (TextUtils.equals(key, key())) {
            data.value = read()
        }
    }

    protected abstract fun key(): String

    protected abstract fun clazz(): Class<T>

    private inner class PreferenceLiveData : MutableLiveData<T>() {
        override fun onActive() {
            super.onActive()
            preferences.registerOnSharedPreferenceChangeListener { sharedPreferences, key ->
                this@PersistentSetting.onSharedPreferenceChanged(sharedPreferences, key)
            }
        }

        override fun onInactive() {
            super.onInactive()
            preferences.registerOnSharedPreferenceChangeListener { sharedPreferences, key ->
                this@PersistentSetting.onSharedPreferenceChanged(sharedPreferences, key)
            }
        }
    }
}
